### PR TITLE
chore(pipelined): Make mypy green

### PR DESCRIPTION
**Pull request by sebathomas**
## Summary

Fix various type issues for pipelined.
- Correct a few wrong argument or variable types
- Add type stubs for mixins
- Ignore a false positive
- Raise ValueError instead of NPE
- Extract method to handle all types of error in tc_ops_pyroute2
- Add null checks in test code

## Test Plan

```
cd $MAGMA_ROOT/lte/gateway
mypy --ignore-missing-imports python/magma/pipelined/
```
```
Success: no issues found in 131 source files
```

## Additional Information

- [ ] This change is backwards-breaking

### CODE REVIEW

File: lte/gateway/python/magma/pipelined/encoding.py [hunk 1]
```diff
@@ -137,7 +137,7 @@ def get_hash(s, hash_function) -> bytes:
 
 def encode_str(s: str, encoding_type) -> str:
     if encoding_type == PipelineD.HEConfig.BASE64:
-        s = codecs.encode(codecs.decode(s, 'hex'), 'base64').decode()
+        s = codecs.encode(codecs.decode(s, 'hex'), 'base64').decode()  # type: ignore
```
Comment (sebathomas) [inline lte/gateway/python/magma/pipelined/encoding.py:140]: This seems to be an issue in mypy, see: https://github.com/python/typeshed/issues/300

File: lte/gateway/python/magma/pipelined/app/BUILD.bazel [hunk 2]
```diff
@@ -204,6 +204,9 @@ py_library(
 py_library(
     name = "restart_mixin",
     srcs = ["restart_mixin.py"],
+    deps = [
```
Comment (nstng) [inline lte/gateway/python/magma/pipelined/app/BUILD.bazel]: NIT for one liner lists we usually do not indent, i.e., `deps = [":startup_flows"],`

Comment (nstng) [inline lte/gateway/python/magma/pipelined/app/BUILD.bazel]: Curious, why was this not needed before? I see that you added the import to `restart_mixin`, but the `_startup_flow_controller` was there before, right?

Comment (LKreutzer) [reply lte/gateway/python/magma/pipelined/app/BUILD.bazel]: The import `from magma.pipelined.app.startup_flows import StartupFlows` was newly added

Comment (nstng) [reply lte/gateway/python/magma/pipelined/app/BUILD.bazel]: > I see that you added the import to `restart_mixin`

My question was, why the import was not needed before.

Comment (LKreutzer) [reply lte/gateway/python/magma/pipelined/app/BUILD.bazel]: To me it looks like both lines are new?

Comment (LKreutzer) [reply lte/gateway/python/magma/pipelined/app/BUILD.bazel]: Current master does this: Screenshot from 2022-06-10 13-05-01

Comment (nstng) [reply lte/gateway/python/magma/pipelined/app/BUILD.bazel]: > but the `_startup_flow_controller` was there before

yes, but the object that contains the `StartupFlows` was existing before

Comment (sebathomas) [reply lte/gateway/python/magma/pipelined/app/BUILD.bazel]: Yes but it wasn't typed here before so the Python code didn't know what it was in the abstract superclass.

**Review by nstng — APPROVED** [review]
lgtm (feel free to ignore the nit if you do not plan to do other changes). sudo tests locally green.

File: lte/gateway/python/magma/pipelined/ebpf/ebpf_manager.py [hunk 3]
```diff
@@ -11,6 +11,8 @@
 limitations under the License.
 """
 
+from __future__ import annotations
```
Comment (voisey) [inline lte/gateway/python/magma/pipelined/ebpf/ebpf_manager.py:14]: What effect does this have here?

Comment (sebathomas) [reply lte/gateway/python/magma/pipelined/ebpf/ebpf_manager.py:14]: In this case, it allows you to use `Array[ctypes.c_ubyte]` as a type. In general, it makes type hints behave in this file like they will in Python 3.10:
https://peps.python.org/pep-0563/

Comment (voisey) [reply lte/gateway/python/magma/pipelined/ebpf/ebpf_manager.py:14]: Makes sense, thanks!

File: lte/gateway/python/magma/pipelined/app/restart_mixin.py [hunk 4]
```diff
@@ -28,6 +35,17 @@ class RestartMixin(metaclass=ABCMeta):
...
+    logger: Logger
+    tbl_num: int
+    cleanup_state: Callable
+    delete_all_flows: Callable
+    finish_init: Callable
+    _msg_hub: MessageHub
+    _datapath: Datapath
+    _startup_flow_controller: Optional[StartupFlows]
+    _startup_flows_fut: Future[StartupFlows]
+    _clean_restart: Callable
```
Comment (voisey) [inline lte/gateway/python/magma/pipelined/app/restart_mixin.py]: Is `_clean_restart` definitely callable? It seems to me that it usually gets set to the `clean_restart` part of the config, which is a boolean.

Comment (sebathomas) [reply lte/gateway/python/magma/pipelined/app/restart_mixin.py]: Good catch, that's an error.

(And it's bad that mypy didn't notice...)

**Review by voisey — APPROVED** [review]

**Pull request merged**